src/lsm/table_iterator.zig

```diff
@@ -90,8 +91,6 @@ pub fn TableIteratorType(comptime Table: type, comptime Storage: type) type {
         }
 
         pub fn deinit(it: *TableIterator, allocator: mem.Allocator) void {
-            assert(!it.read_pending);
```

Comment:
Was this not necessary? (Or maybe it makes it hard to simulate a crash and restart if deinit panics?)

Reply:
> (Or maybe it makes it hard to simulate a crash and restart if deinit panics?)

That's it exactly. We deinit the replica (and all the way down) when simulating a crash — which can occur at any time.

Reply:
Is it possible that that would result in the io layer delivering a read to a freed table?

Reply:
For these purposes, we would be using the test storage (`src/test/storage.zig`) which is running in our same thread. When we "crash" a replica, we call `Storage.reset()`, which clears the pending reads/writes (and as of this PR, injects faults into some of the write targets). If it did let the write finish that would be ok too, though.

Reply:
For crashes, would it be easier to just deinit the arena allocator? After all, we won't call deinit in real crashes either.

Reply:
We don't actually use an arena allocator within the VOPR. But even if we did, having this extra granularity is useful:

- e.g. it lets us reuse the message pool instead of freeing/allocating it, which saves some time. (The message pool is a whole bunch of 1MB allocations).
- we can test `deinit()` functions — even though the real replica doesn't ever use it, it is still nice to have.

Reply:
The crash logic is here by the way: https://github.com/tigerbeetledb/tigerbeetle/blob/55cdb208814aa9ac49ec7521c428653eda347392/src/test/cluster.zig#L281-L321

It is simplified a little in this PR (around counting packets in the network). Most of the complexity is around determining whether or not we can safely crash a replica, to ensure that VSR recovery protocol will be able to bring it back.

It will be dramatically simplified within the next month or so when we switch to using the superblock for recovery instead.